llvm/lib/CASObjectFormats/NestedV1.cpp

```diff
@@ -1853,9 +1857,14 @@ Error LinkGraphBuilder::makeSymbol(Expected<SymbolRef> Symbol) {
   //
   // FIXME: Handle abstract backedges somehow more directly (likely as part of
   // the above).
-  while (!Worklist.empty())
-    if (Error E = defineSymbol(Worklist.pop_back_val()))
+  while (!Worklist.empty()) {
+    SymbolToDefine Sym = Worklist.pop_back_val();
+    // Insert sub-symbols in reverse order (from same insertion point) so when
+    // they get popped they will get processed in original order.
```

Comment:
This comment is misleading since this is not guaranteed to match the original order. The symbols were sorted in various ways before serializing and there's no going back. This is intentional, so that:
```
int f1() { return 1; }
int f2() { return 2; }
```
gets the same CAS object as
```
int f2() { return 2; }
int f1() { return 1; }
```
(See also the `FIXME` above about block order. Perhaps that needs to be updated?)

Inline and templates (and similar) are stored in a DAG by call graph. This worklist managing doing a DFS through the DAG. It looks to me like both before and after this patch, the blocks/symbols will be added in pre-order (first the caller, then the children). The difference is the order of the children (and thus their subgraphs). Can you confirm?

I think we want to try to match what Clang IRGen does, which should be in order of edge offsets. Can you confirm that's what will be happening now?

Reply:
> The difference is the order of the children (and thus their subgraphs). Can you confirm?

That is correct.

> I think we want to try to match what Clang IRGen does, which should be in order of edge offsets. Can you confirm that's what will be happening now?

As long as new symbols/blocks are created for children edges, they will be in-order, AFAICT.

Reply:
Okay, sounds good.
> (See also the FIXME above about block order. Perhaps that needs to be updated?)
Please have a look at the FIXME about block order. This patch is reifying the current approach of pre-order, but the FIXME says that we want post-order.

The main benefit of post-order is as a hack to fix order of CIE/FDE in EH frames. CIEs are like abbreviations, and must come before any FDE that references it (every FDE references exactly one CIE). Maybe you can fine-tune the FIXME to be more specific about that, and clarify somewhere (FIXME or otherwise) that the current order (pre-order with children in order) is intended/desirable outside of special cases like EH frame.